chore: enhance documentation of the rules engine (#1736)
AmadeusITGroup/otter: docs/rules-engine/README.md

# Otter Rules Engine

## Introduction

This document describes the various concepts and models that are part of the [Otter rules engine module](https://www.npmjs.com/package/@o3r/rules-engine).
It serves as an introduction course before you can integrate it within your application following the
[rules engine implementation guide](./how-to-use/README.md).

> As a reminder, the rules engine is a client-based mechanism that evaluates a list of [Rulesets](#ruleset) to define [actions](#action) to

Reviewer: The `As a reminder` looks weird to me in the document that is supposed to introduce the concept
```suggestion
In a nutshell, the rules engine is a client-based mechanism that evaluates a list of [Rulesets](#ruleset) to define [actions](#action) to
```

Author: It is a reminder on concepts covered before in other documentation (the npmjs), so in a nutshell would not fit.
Maybe, 
```suggestion
As covered in the [package documentation](https://www.npmjs.com/package/@o3r/rules-engine), the rules engine is a client-based mechanism that evaluates a list of [Rulesets](#ruleset) to define [actions](#action) to
```
What do you think?

> This ensures you only import packages you actually need.

### Fact

A fact is a stream of values.

During a rule execution, its conditions will be evaluated using the latest fact value(s).
Every fact can be identified by its `id` property and exposes a `value$` stream that will trigger a new evaluation
of the Ruleset it has been linked for every value it emits.
This allows a big optimization since the rules engine will react to a fact change in order to execute only the subset
of rules that relies on this fact.

Facts can be defined inside the code of the application, as described in the [custom fact documentation](./how-to-use/custom-fact.md),
or can be defined by the ruleset itself. 

In that case, we call it a [runtime fact](#runtime-fact).

> [!WARNING]
> A rule depending on a fact will be re-evaluated whenever this fact is updated. 
> If you use a fact that will be updated a lot in a short time frame, you may encounter performance issues.
> 
> You can avoid this by:
> * disabling the ruleset if the component impacted by the rule is missing from the page (see [linkedComponents](./how-to-use/dedicated-component-ruleset.md))

> > * avoiding using [JSONPath](./examples/complex-fact.md) on facts that are updated a lot in a short time frame and

Reviewer 2: should now be less a problem with the distinctUntilChanged : 
![image](https://github.com/AmadeusITGroup/otter/assets/122815580/31c19b3e-befa-4643-b0a1-61be091fd9c8)

Reviewer 2: But still a good practice, and if a subset is used several times it's better to create a dedicated fact instead of applying the same jsonPath several times on the same fact.

Author: As discussed, I will remove it as it is sensibly the same (JSONPath versus map / pick operator that will be evaluated as many times)

### Rule

A ``Rule`` contains a set of conditions that outputs a list of actions after processing. 

These conditions and actions are referred to as blocks and follow a tree architecture: 
a block can be an ``IfElseBlock`` (branch of the tree) or an ``ActionBlock`` (leave of the tree).

The ``IfElseBlock`` contains conditions to meet as well as the blocks to execute depending on the success or failure of 
the condition. 
The conditions can be linked via ``AND`` and ``OR`` expressions.

An ``IfElseBlock`` can result in another ``IfElseBlock`` or an ``ActionBlock``. 

This allows for nested conditions as demonstrated by the [nested condition Ruleset example](./examples/nested-conditions.md).

> [!NOTE]
> If no condition is defined in the ``IfElseBlock``, its ``successElements`` will always be executed.

A ``Rule`` also contains information needed to optimize the reevaluation of its conditions:
* ``inputRuntimeFacts`` : the input runtime facts the rule is based on
* ``inputFacts`` : the application facts that are used by the rule
* ``outputRuntimeFacts`` : the runtime facts that are updated/set by the rule

> Thanks to these parameters, the rule engines can make sure to only trigger rules evaluation if the input and output of

Reviewer 3: ```suggestion
Thanks to these parameters, the rule engines can make sure to only trigger rules evaluation if the `inputFacts` of
```
The `outputRuntimeFacts` cannot be the trigger of the rule evaluation, as they depends on `inputFacts` no ?

> (such as a path `/path/to/file`) should be preceded by a double backslash `\\` (for example `\\/path/to/file`)

### Action

``ActionBlock`` describes the action to execute. Each ``Action`` is identified
by a unique id to identify the process to trigger and an eventual payload.

The Otter framework provides a few build-in actions linked to Otter specific features:
* ``UPDATE_CONFIG`` modifies the configuration of a configurable component
(provided by [@o3r/configuration](https://www.npmjs.com/package/@o3r/configuration))
* ``UPDATE_ASSET`` overrides the path of an asset which uses the ``o3rDynamicContent`` pipe 
(provided by [@o3r/dynamic-content](https://www.npmjs.com/package/@o3r/dynamic-content))
* ``UPDATE_LOCALISATION`` substitutes the translation of a key with the one from another key.
([@o3r/localization](https://www.npmjs.com/package/@o3r/localization))
* ``UPDATE_PLACEHOLDER`` injects html code in a specific placeholder
([@o3r/components](https://www.npmjs.com/package/@o3r/components))

You can provide your own custom action following [this guide](./how-to-use/custom-action.md).

> [!NOTE]

> > The actions handled by the rules engine need to be [registered](./how-to-use/README.md#step-2-register-the-action-handlers-in-the-runner) in your application to be recognized by the service.

Reviewer 3: ```suggestion
> The actions handled by the rules engine need to be [registered](./how-to-use/README.md#step-2-register-the-action-handlers-in-the-runner) in your application to be recognized by the engine.
```

> This ensures you only import packages you actually needs.

> ### Fact

Reviewer 3: Maybe we should have the Fact paragraph before the Rule one because we use several time the "fact" concept in the Rule section

A fact is a stream of values.

During a rule execution, its conditions will be evaluated using the latest fact value(s).
Every fact can be identified by its ``id`` property and exposes a ``value$`` stream that will trigger a new evaluation
of the Ruleset it has been linked for every value it emits.
This allows a big optimization since the rules engine will react to a fact change in order to execute only the subset
of rules that relies on this fact.

Facts can be defined inside the code of the application as the example below, or can be defined by the ruleset itself.
In that case, we call it a [runtime fact](#runtime-fact).

> ```typescript

Reviewer 3: I'm not fan of this kind of example as it will not be extracted like this
